docs/workflow-lifecycle.md

```diff
@@ -0,0 +1,239 @@
+Frontend->>User Application: Start Succeed
+loop QueueProcessor
+    History->>Persistence: GetHistoryTasks
+		History->>History: ProcessTask
+		History->>Matching: AddWorkflowTask<br>
+end
+```
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- History service [`StartWorkflow` handler](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/api/startworkflow/api.go#L157).
+- History service [queue processors](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/history_engine.go#L303) and [transfer task queue processor](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/queues/queue_immediate.go#L150).
+
+</details>
+<br>
+
+---
+
+<br>
+
+2. **The Worker dequeues and processes the Workflow Task**
+   - It advances the workflow execution, and becomes blocked on the Activity call.
+
+```mermaid
+sequenceDiagram
+Worker->>Frontend: PollWorkflowTask
+Frontend->>Matching: PollWorkflowTask
+History->>Matching: AddWorkflowTask (step 1 above)
+Matching->>History: RecordWorkflowTaskStarted
+History->>Persistence: UpdateWorkflowExecution
+note over Persistence: append History Event: WorkflowTaskStarted<br>update Mutable State & add Timer Task (workflow task timeout)
+Persistence->>History: Update Succeed
+History->>Matching: Record Succeed
+Matching->>Frontend: WorkflowTask
+Frontend->>Persistence: GetHistoryEvents
+Persistence->>Frontend: History Events
+Frontend->>Worker: WorkflowTask
+Worker->>Worker: Advance workflow
+```
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- History service [`RecordWorkflowTaskStarted` handler](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/handler.go#L319)
+
+</details>
+<br>
+
+---
+
+<br>
+
+3. **The Worker sends a `ScheduleActivityTask` command**
+   - An Activity task is added in the Matching service.
+
+```mermaid
+sequenceDiagram
+Worker ->> Frontend: RespondWorkflowTaskCompleted<br>[ScheduleActivityTask]
+Frontend->> History: RespondWorkflowTaskCompleted<br>[ScheduleActivityTask]
+History ->> Persistence: UpdateWorkflowExecution
+note over Persistence: append History Events: WorkflowTaskCompleted, ActivityTaskScheduled<br>update MutableState and add Transfer Task
+Persistence ->> History: Update Succeed
+History->>Frontend: Respond Succeed
+Frontend->>Worker: Respond Succeed
+loop QueueProcessor
+    History->>Persistence: GetHistoryTasks
+		History->>History: ProcessTask
+		History->>Matching: AddActivityTask
+end
+```
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- History service [`ScheduleActivityTask` command handler](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/workflow_task_handler.go#L338)
+
+</details>
+<br>
+
+---
+
+<br>
+
+**4. The Worker dequeues the Activity task and executes the activity**
```

Comment:
Similarly, we're making an assumption here that the activity is running on the same worker as the workflow (which is a good assumption to make, but we might want to mention it here):

```suggestion
**4. The Worker dequeues the Activity task and executes the activity.**
  - (Note: This Worker may or may not be the same worker that launched the activity, depending on what other workers are available, the type of the activity, etc.)
```

Reply:
Good call. I've changed it to "A Worker".